Use integrated wasm-opt

brson opened the pull request:
Compile `wasm-opt` and call it as a library instead of as an external binary.

This PR has three patches that result in completely removing support for the `wasm-opt` CLI in favor of the crate. If desired, the third commit can be excluded, leaving support for both methods.

Documentation for the `wasm-opt` bindings:

- https://github.com/brson/wasm-opt-rs
- https://docs.rs/wasm-opt/latest/wasm_opt/

Things to be aware of about the `wasm-opt` crate:

- It depends on Rust 1.48+
- It depends on a c++17 compiler, and should emit a readable error if one is not available.
- The `wasm-opt-sys` crate takes a non-negligible amount of time to build (minutes on my underpowered laptop). It also does not do any incremental recompilation, so if the build is invalidated it will rebuild the C++ code from scratch. The lack of incremental recompilation is a limitation self-imposed by not using cmake or other external build system.
- `wasm-opt` on windows does not support extended unicode paths (probably anything non-ascii, maybe latin-1, it's unclear). This is a limitation of binaryen and not a regression of the bindings. It may or may not be fixed in the future. The API will return an error if this occurs.
- `cargo tarpaulin` (code coverage) segfaults running any `wasm-opt` crates (https://github.com/brson/wasm-opt-rs/issues/59), reason unknown. This behavior could infect other crates that link to `wasm-opt`. If you use tarpaulin, you might check it.

Fixes https://github.com/paritytech/cargo-contract/issues/733

ascjones reviewed (approved):
This is absolutely fantastic thank you @brson, it removes some major friction from our installation process!

> It depends on a c++17 compiler, and should emit a readable error if one is not available.

I wonder whether we should include that as a prerequisite. Which reminds me, we should remove `Step 2: install binaryen` from https://github.com/paritytech/cargo-contract#installation

athei commented:
We should update the README in here, too? Need to remove `wasm-opt` from the list of required dependencies.

brson commented:
I have updated the readme to remove the binaryen installation step, and added an entry to the changelog.

I am not sure where in the readme to mention the c++17 prerequisite.

brson commented:
I've also added a mention of c++17 to the installation section.

cmichi commented on crates/cargo-contract/src/main.rs:
```diff
@@ -71,6 +72,12 @@ use assert_cmd as _;
+use regex as _;
+
+// Only used on windows.
+use which as _;
```
Is it not possible to prefix with `#[cfg(target_os = "windows")]` here?

brson replied:
I don't think so. This is intentionally doing the opposite - making it look like `which` is used on not-windows. `cargo-contract` has a clippy lint turned on to check that linked crates are actually used, and `which` is only used on windows, so clippy doesn't see it on unix. What could be done is to change the `Cargo.toml` to only link to `which` on windows.

cmichi commented:
This is a great achievement, that hopefully will also benefit the broader Rust+Wasm ecosystem. 

Thanks a lot, @brson!
